src/lib/components/documents/ResultsList.svelte

```diff
@@ -1,37 +1,100 @@
 <script context="module" lang="ts">
   import { writable, type Writable } from "svelte/store";
+  import Button from "../common/Button.svelte";
 
-  export const selected: Writable<(number | string)[]> = writable([]);
+  // IDs might be strings or numbers, depending on the API endpoint
+  // enforce type consistency here to avoid comparison bugs later
+  export const selected: Writable<string[]> = writable([]);
+  export let visible: Writable<Set<string>> = writable(new Set());
+
+  export let total: Writable<number> = writable(0);
 </script>
 
 <script lang="ts">
+  import type { Document, DocumentResults } from "$lib/api/types";
+
+  import { onMount } from "svelte";
   import { _ } from "svelte-i18n";
-  import type { DocumentResults } from "$lib/api/types";
+
   import DocumentListItem from "./DocumentListItem.svelte";
   import Flex from "../common/Flex.svelte";
-  import Checkbox from "../common/Checkbox.svelte";
   import { Search24 } from "svelte-octicons";
   import Empty from "../common/Empty.svelte";
 
-  export let results: DocumentResults = undefined;
+  export let results: Document[] = [];
+  export let count: number = undefined;
+  export let next: string | null = null;
+  export let auto = false;
+
+  let loading = false;
+  let end: HTMLElement;
+  let observer: IntersectionObserver;
+
+  // track what's visible so we can compare to $selected
+  $: $visible = new Set(results.map((d) => String(d.id)));
 
-  function updateSelection(event: Event, id: string | number) {
-    const { checked } = event.target as HTMLInputElement;
-    if (checked) {
-      selected.set([...$selected, id]);
-    } else {
-      selected.set([...$selected.filter((d) => d !== id)]);
+  // load the next set of results
+  async function load(url: URL) {
+    loading = true;
+    const res = await fetch(url, { credentials: "include" });
+
+    if (!res.ok) {
+      // todo: better error handling
+      console.error(res.statusText);
+      loading = false;
     }
+
+    const r: DocumentResults = await res.json();
+
+    results = [...results, ...r.results];
+    $total = r.count;
+    next = r.next;
+    loading = false;
+    if (auto) watch(end);
+  }
+
+  function watch(el: HTMLElement) {
+    const io = new IntersectionObserver((entries, observer) => {
+      entries.forEach(async (entry) => {
+        if (entry.isIntersecting && next) {
+          await load(new URL(next));
+          observer.unobserve(el);
+        }
+      });
+    });
+
+    io.observe(el);
+    return io;
+  }
+
+  function unwatch(io: IntersectionObserver, el: HTMLElement) {
+    io.unobserve(el);
   }
```

Comment:
`watch` and `unwatch` are great code, I think they could be extracted into a very useful util:

```
  function watch(el: HTMLElement, callback: (entry, observer, el) => void) {
    const io = new IntersectionObserver((entries, observer) => {
      entries.forEach(async (entry) => callback(entry, observer, el));
    });

    io.observe(el);
    return io;
  }

  function unwatch(io: IntersectionObserver, el: HTMLElement) {
    io.unobserve(el);
  }
```
and then called here as:
```
function onIntersect(entry, observer, el) {
  if (entry.isIntersecting && next) {
    await load(new URL(next));
    observer.unobserve(el);
  }
}

watch(end, onIntersect)
```

Reply:
Yeah, happy to consolidate this into a util next time we need it. I'm avoiding creating utilities until we know we'll need them in more than one place.